### PR TITLE
PH46816: Header validation

### LINKED ISSUE #22425: IBM WebSphere Application Server and IBM WebSphere Application Server Liberty are vulnerable to HTTP header injection (CVE-2022-34165 CVSS 5.4)
**Describe the bug**  
IBM WebSphere Application Server and IBM WebSphere Application Server Liberty are vulnerable to HTTP header injection (CVE-2022-34165 CVSS 5.4)

**Steps to Reproduce**  
n/a

**Expected behavior**  
n/a

**Diagnostic information:**  
 - OpenLiberty Version:  all versions

**Additional context**  
n/a

### PR DESCRIPTION
**Pull request by loriadi**
fixes #22425

### CODE REVIEW

**Comment by loriadi** [comment]
fixes #PH46816

**Comment by loriadi** [comment]
#build

**Review by mrsaldana — APPROVED** [review]

**Comment by loriadi** [comment]
#build

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/internal/BNFHeadersImpl.java [hunk 1]
```diff
@@ -4497,6 +4525,69 @@ private void processForwardedAddressExtract(String nodeExtract, ListType type) {
...
     }
 
+    /*
+     * A valid proto may start with an alpha followed by any number of chars that are
+     * - alpha
+     * - numeric
+     * - "+" or "-" or "."
+     */
+
+    private boolean validateProto(String forwardedProto) {
```
Comment (tevans78) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/internal/BNFHeadersImpl.java:4535]: These two new methods look like they would have benefited from being unit tested? Suggest that you open a follow up PR to add that if possible.

File: dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java [hunk 2]
```diff
@@ -920,16 +920,56 @@ public String getRemoteHostAddress() {
...
+            remoteAddr = contextRemoteHostAddress();
+        }
 
         return remoteAddr;
     }
 
+    /*
+     * Validate the remote address
+     */
+    boolean validateRemoteAddress(String address) {
```
Comment (tevans78) [inline dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java:940]: Again, could this have been unit tested?

File: dev/com.ibm.ws.webcontainer.servlet.3.1_fat/publish/servers/servlet31_privateHeaderServer/bootstrap.properties [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+# are made available under the terms of the Eclipse Public License v1.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-v10.html
+#
+# Contributors:
+#     IBM Corporation - initial API and implementation
+###############################################################################
+bootstrap.include=../testports.properties
+osgi.console=7777
+com.ibm.ws.logging.trace.specification=*=info:com.ibm.ws.webcontainer*=all:com.ibm.wsspi.webcontainer*=all:HttpTransport=all:HTTPChannel=all:TCPChannel=all:GenericBNF=all:com.ibm.ws.jsp*=all:com.ibm.ws.session.*=all:com.ibm.ws.runtime.update.internal*=all:ChannelFramework=all
```
Comment (tevans78) [inline dev/com.ibm.ws.webcontainer.servlet.3.1_fat/publish/servers/servlet31_privateHeaderServer/bootstrap.properties:13]: This looks like a lot of trace to be switched on by default?

File: dev/com.ibm.ws.webcontainer.servlet.3.1_fat/publish/servers/servlet31_privateHeaderServer/server.xml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+    <featureManager>
+        <feature>servlet-3.1</feature>
+        <feature>jndi-1.0</feature>
+    </featureManager>
+    
+    <httpDispatcher trustedHeaderOrigin="*" trustedSensitiveHeaderOrigin="*"/>
+
+    <logging
+        traceSpecification="*=info=enabled:com.ibm.ws.webcontainer*=all:com.ibm.wsspi.webcontainer*=all:HttpTransport=all:HTTPChannel=all:TCPChannel=all:GenericBNF=all:com.ibm.ws.jsp*=all:com.ibm.ws.session.*=all:com.ibm.ws.runtime.update.internal*=all:ChannelFramework=all"
```
Comment (tevans78) [inline dev/com.ibm.ws.webcontainer.servlet.3.1_fat/publish/servers/servlet31_privateHeaderServer/server.xml:23]: probably best to only define the trace string in one place?

**Comment by loriadi** [comment]
#libby-bot

File: dev/com.ibm.ws.webcontainer.servlet.3.1_fat/publish/servers/servlet31_privateHeaderServer/server.xml (hunk 4 above)
Comment (loriadi) [reply dev/com.ibm.ws.webcontainer.servlet.3.1_fat/publish/servers/servlet31_privateHeaderServer/server.xml:23]: Issue 22470: pared down the trace in serverl.xml

File: dev/com.ibm.ws.webcontainer.servlet.3.1_fat/publish/servers/servlet31_privateHeaderServer/bootstrap.properties (hunk 3 above)
Comment (loriadi) [reply dev/com.ibm.ws.webcontainer.servlet.3.1_fat/publish/servers/servlet31_privateHeaderServer/bootstrap.properties:13]: Issue 22470: removed this trace string and updated the trace string in server.xml

**Pull request merged**